qtt/instrument_drivers/virtualAwg/awgs/ZurichInstrumentsHDAWG8.py

```diff
@@ -66,94 +74,37 @@ def retrieve_running_mode(self):
         raise NotImplementedError
 
     def update_sampling_rate(self, sampling_rate):
-        raise NotImplementedError
+        for key, value in self.__sampling_rate_map.items():
+            if sampling_rate == value:
+                self.__awg.set('awgs_{}_time'.format(self.__awg_number), key)
+                return
+        raise ValueError('Sampling rate {} not in available a list of available values: {}'.format(
+            sampling_rate, self.__sampling_rate_map))
 
     def retrieve_sampling_rate(self):
-        #ziDAQ('setInt', '/dev8044/awgs/0/time', 0);
-        return 2.4e9
+        sample_rate = self.__awg.get('awgs_{}_time'.format(self.__awg_number))
+        return self.__sampling_rate_map[sample_rate]
 
     def update_gain(self, gain):
-        raise NotImplementedError
+        [self.__awg.set('sigouts_{}_range'.format(ch), gain) for ch in self._channel_numbers]
 
     def retrieve_gain(self):
-        # ziDAQ('setDouble', '/dev8044/sigouts/0/range', 1);
-        return 1
+        gains = [self.__awg.get('sigouts_{}_range'.format(ch)) for ch in self._channel_numbers]
+        if not all(g == gains[0] for g in gains):
+            raise ValueError('Not all channel amplitudes are equal. Please reset!')
```

Comment:
The function name has `retrieve`, the call to `__awg` has `range` and the exception raised `amplitude`. There can be only one! The retrieve is mandated by the `AwgCommon`, but at least rename the amplitudes to either range or gain.
Or do you really mean amplitude?